### PR TITLE
*: Convert RegexpExpr to ScalarFuntion and check the arguments length of builtin functions

### PR DESCRIPTION
**Pull request by zimulala**

### CODE REVIEW

**Comment by zimulala** [comment]
PTAL @hanfei19910905 @coocood @shenli

File: expression/expression.go [hunk 1]
```diff
@@ -323,8 +327,9 @@ func ComposeCNFCondition(conditions []Expression) Expression {
 	if length == 1 {
 		return conditions[0]
 	}
-	return NewFunction(ast.AndAnd,
+	expr, _ := NewFunction(ast.AndAnd,
```
Comment (shenli) [inline expression/expression.go]: Why ignore the err?

File: plan/expression_rewriter.go [hunk 2]
```diff
@@ -465,32 +476,56 @@ func (er *expressionRewriter) rewriteVariable(v *ast.VariableExpr) bool {
...
 
 func (er *expressionRewriter) notToScalarFunc(b bool, op string, tp *types.FieldType,
 	args ...expression.Expression) *expression.ScalarFunction {
-	opFunc := expression.NewFunction(op, tp, args...)
+	opFunc, err := expression.NewFunction(op, tp, args...)
+	if err != nil {
+		er.err = errors.Trace(err)
+		return nil
+	}
 	if !b {
```
Comment (shenli) [inline plan/expression_rewriter.go]: b is not clear enough. Suggest to change it to hasNot.

File: executor/executor_test.go [hunk 3]
```diff
@@ -1415,6 +1406,51 @@ func (s *testSuite) TestBuiltin(c *C) {
...
+	testCases = []testCase{
+		{"^$", "a", 0},
+		{"a", "a", 1},
+		{"a", "b", 0},
+		{"aA", "aA", 1},
+		{".", "a", 1},
+		{"^.$", "ab", 0},
+		{"..", "b", 0},
+		{".ab", "aab", 1},
+		{".*", "abcd", 1},
```
Comment (shenli) [inline executor/executor_test.go]: Add a case that pattern is partial match val. For example: "ab." regexp "abcddd".

File: evaluator/builtin_other.go [hunk 4]
```diff
@@ -132,6 +134,28 @@ func builtinLike(args []types.Datum, _ context.Context) (d types.Datum, err erro
 	return
 }
 
+func builtinRegexp(args []types.Datum, _ context.Context) (d types.Datum, err error) {
```
Comment (shenli) [inline evaluator/builtin_other.go]: Add reference: http://dev.mysql.com/doc/refman/5.7/en/regexp.html#operator_regexp

File: expression/expression.go (hunk 1 above)
Comment (zimulala) [inline expression/expression.go]: Yep. Because here error is always nil.

**Comment by shenli** [comment]
LGTM

File: evaluator/builtin.go [hunk 5]
```diff
@@ -149,7 +149,8 @@ var Funcs = map[string]Func{
 	ast.In:         {builtinIn, 1, -1},
 	ast.IsTruth:    {isTrueOpFactory(opcode.IsTruth), 1, 1},
 	ast.IsFalsity:  {isTrueOpFactory(opcode.IsFalsity), 1, 1},
-	ast.Like:       {builtinLike, 1, 3},
+	ast.Like:       {builtinLike, 5, 5},
```
Comment (hanfei19910905) [inline evaluator/builtin.go]: why should like have five args?

File: plan/expression_rewriter.go [hunk 6]
```diff
@@ -550,30 +593,64 @@ func (er *expressionRewriter) likeToScalarFunc(v *ast.PatternLikeExpr) {
 	er.ctxStack = append(er.ctxStack, function)
 }
 
+func (er *expressionRewriter) regexpToScalarFunc(v *ast.PatternRegexpExpr) {
+	l := len(er.ctxStack)
+	function := er.notToScalarFunc(v.Not, ast.Regexp, v.Type, er.ctxStack[l-2], er.ctxStack[l-1],
+		&expression.Constant{Value: types.NewDatum(v.Re)}, &expression.Constant{Value: types.NewDatum(v.Sexpr)})
```
Comment (coocood) [inline plan/expression_rewriter.go]: `v.Re` and `v.Sexpr` is generated value, we don't need to include them.

Comment (coocood) [inline plan/expression_rewriter.go]: So does `v.PatChars' and 'v.PatTypes' in Like expression.

**Comment by zimulala** [comment]
PTAL @coocood @hanfei19910905

**Comment by coocood** [comment]
LGTM

File: plan/expression_rewriter.go [hunk 7]
```diff
@@ -535,45 +574,80 @@ func (er *expressionRewriter) caseToScalarFunc(v *ast.CaseExpr) {
...
 	er.ctxStack = append(er.ctxStack, function)
 }
 
 func (er *expressionRewriter) funcCallToScalarFunc(v *ast.FuncCallExpr) {
 	l := len(er.ctxStack)
 	function := &expression.ScalarFunction{FuncName: v.FnName}
-	for i := l - len(v.Args); i < l; i++ {
-		function.Args = append(function.Args, er.ctxStack[i])
-	}
+	function.Args = er.ctxStack[l-len(v.Args)-1:]
```
Comment (hanfei19910905) [inline plan/expression_rewriter.go]: Why l -len(v.Args)-1 but l-len(v.Args)

**Comment by hanfei19910905** [comment]
LGTM

**Pull request merged**